Mathlib/Analysis/NormedSpace/HahnBanach/SeparatingDual.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2023 Sébastien Gouëzel All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+import Mathlib.Analysis.NormedSpace.HahnBanach.Extension
+import Mathlib.Analysis.NormedSpace.HahnBanach.Separation
+
+/-!
+# Spaces with separating dual
+
+We introduce a typeclass `SeparatingDual R V`, registering that the points of the topological
+module `V` over `R` can be separated by continuous linear forms.
+
+This property is satisfied for normed spaces over `ℝ` or `ℂ` (by the analytic Hahn-Banach theorem)
+and for locally convex topological spaces over `ℝ` (by the geometric Hahn-Banach theorem).
+
+Under the assumption `SeparatingDual R V`, we show in
+`SeparatingDual.exists_continuousLinearMap_apply_eq` that the group of continuous linear
+equivalences acts transitively on the set of nonzero vectors.
+-/
+
+/-- When `E` is a topological module over a topological ring `R`, the class `SeparatingDual R E`
+registers that continuous linear forms on `E` separate points of `E`. -/
+class SeparatingDual (R V : Type*) [Ring R] [AddCommGroup V] [TopologicalSpace V]
+    [TopologicalSpace R] [Module R V] : Prop :=
+  /-- Any nonzero vector can be mapped by a continuous linear map to a nonzero scalar. -/
+  exists_ne_zero' : ∀ (x : V), x ≠ 0 → ∃ f : V →L[R] R, f x ≠ 0
+
+instance {E : Type*} [TopologicalSpace E] [AddCommGroup E] [TopologicalAddGroup E]
+    [Module ℝ E] [ContinuousSMul ℝ E] [LocallyConvexSpace ℝ E] [T1Space E] : SeparatingDual ℝ E :=
+  ⟨fun x hx ↦ by
+    rcases geometric_hahn_banach_point_point hx.symm with ⟨f, hf⟩
+    simp only [map_zero] at hf
+    exact ⟨f, hf.ne'⟩ ⟩
+
+instance {E 𝕜 : Type*} [IsROrC 𝕜] [NormedAddCommGroup E] [NormedSpace 𝕜 E] : SeparatingDual 𝕜 E :=
+  ⟨fun x hx ↦ by
+    rcases exists_dual_vector 𝕜 x hx with ⟨f, -, hf⟩
+    refine ⟨f, ?_⟩
+    simpa [hf] using hx⟩
+
+namespace SeparatingDual
+
+section Ring
+
+variable {R V : Type*} [Ring R] [AddCommGroup V] [TopologicalSpace V]
+  [TopologicalSpace R] [Module R V] [SeparatingDual R V]
+
+lemma exists_ne_zero {x : V} (hx : x ≠ 0) :
+    ∃ f : V →L[R] R, f x ≠ 0 :=
+  exists_ne_zero' x hx
+
+theorem exists_separating_of_ne {x y : V} (h : x ≠ y) :
+    ∃ f : V →L[R] R, f x ≠ f y := by
+  rcases exists_ne_zero (R := R) (sub_ne_zero_of_ne h) with ⟨f, hf⟩
+  exact ⟨f, by simpa [sub_ne_zero] using hf⟩
+
+protected theorem t1Space [T1Space R] : T1Space V := by
+  apply t1Space_iff_exists_open.2 (fun x y hxy ↦ ?_)
+  rcases exists_separating_of_ne (R := R) hxy with ⟨f, hf⟩
+  exact ⟨f ⁻¹' {f y}ᶜ, isOpen_compl_singleton.preimage f.continuous, hf, by simp⟩
+
+protected theorem t2Space [T2Space R] : T2Space V := by
+  apply t2Space_iff_nhds.2 (fun {x} {y} hxy ↦ ?_)
+  rcases exists_separating_of_ne (R := R) hxy with ⟨f, hf⟩
+  rcases t2_separation_nhds hf with ⟨u, v, hu, hv, huv⟩
+  exact ⟨f ⁻¹' u, f.continuous.continuousAt hu, f ⁻¹' v, f.continuous.continuousAt hv,
+    huv.preimage f⟩
```

Comment:
```suggestion
  apply (t2Space_iff _).2 (fun {x} {y} hxy ↦ ?_)
  rcases exists_separating_of_ne (R := R) hxy with ⟨f, hf⟩
  exact separated_by_continuous f.continuous hf
```

Meta: Is it okay to add small suggestions like that on delegated PRs, or is it bad practice?

Reply:
And thanks, by the way!